### PR TITLE
refactor(authentication-service): streamline state handling in KeycloakLoginController and add optional state property to ClientAuthRequest

### PR DESCRIPTION
**Pull request by subhamjagurisf**
## Description

This PR refactors  [PR # 2353](https://github.com/sourcefuse/loopback4-microservice-catalog/pull/2353 ) **flexible state parameter handling** in the Keycloak authentication flow while maintaining **100% backward compatibility** with existing implementations.

##Fixes: https://github.com/sourcefuse/loopback4-microservice-catalog/issues/2352 

### Summary of Changes

Previously, the `state` parameter in the Keycloak OAuth flow was hardcoded to only contain `client_id`, preventing applications from passing additional contextual information through the authentication process. This enhancement allows applications to:

- Pass custom state parameters (e.g., redirect paths, session IDs, tenant information)
- Preserve application context across the OAuth redirect flow
- Implement deep linking after authentication
- Maintain multi-tenant context during login

**Key Features:**

1. ✅ **Backward Compatible** - Existing projects work unchanged (no migration needed)
2. ✅ **Automatic `client_id` handling** - Appends `client_id` automatically if not present in custom state
3. ✅ **State preservation** - All state parameters (except `client_id`) are returned with the authorization code
4. ✅ **OAuth 2.0 compliant** - Follows RFC 6749 recommendations

### Implementation Details

**Modified Files:**

- `services/authentication-service/src/modules/auth/controllers/keycloak-login.controller.ts`
  - Enhanced `queryGen` function to handle three scenarios:
    1. No state passed (default) → `client_id=xyz` (maintains current behavior)
    2. State without `client_id` → Automatically appends `client_id`
    3. State with `client_id` → Uses as-is, preserves all parameters
  - Updated `keycloakCallback` method to parse all state parameters and include them in the redirect URL

**New Files:**

- `services/authentication-service/src/__tests__/unit/keycloak-login.controller.unit.ts`
  - Comprehensive unit test suite (543 lines, 21 test cases)
  - Tests all three state handling scenarios
  - Validates error handling and edge cases
  - Confirms backward compatibility

### Usage Example

**Before (Still Works):**

```javascript
POST /auth/keycloak
{
  "client_id": "my-app",
  "client_secret": "secret"
}
// Redirect: https://app.example.com/callback?code=TOKEN
```

**After (New Capability):**

```javascript
POST /auth/keycloak
{
  "client_id": "my-app",
  "client_secret": "secret",
  "state": "redirect_path=/dashboard/reports&filter=monthly&tenant=acme"
}
// Redirect: https://app.example.com/callback?code=TOKEN&redirect_path=/dashboard/reports&filter=monthly&tenant=acme
```

### Backward Compatibility Verification

| Scenario        | Before                | After                 | Status               |
| --------------- | --------------------- | --------------------- | -------------------- |
| No state passed | `state=client_id=xyz` | `state=client_id=xyz` | ✅ Identical         |
| Redirect URL    | `?code=TOKEN`         | `?code=TOKEN`         | ✅ Identical         |
| Existing apps   | Work as-is            | Work as-is            | ✅ No changes needed |

Fixes #(issue-number)

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] Backward compatible enhancement

### Unit Tests

- [x] **21 comprehensive unit tests** covering all scenarios (all passing in 59ms)
  - 9 tests for `queryGen` function (state handling logic)
  - 12 tests for `keycloakCallback` method (redirect with state parameters)

### Test Scenarios Covered:

- [x] **Scenario 1:** No state passed (default behavior - backward compatibility)
- [x] **Scenario 2:** State passed without `client_id` (automatic appending)
- [x] **Scenario 3:** State passed with `client_id` and other properties (preservation)
- [x] **Error scenarios:** Missing `client_id`, undefined user, client not found, no redirect URL
- [x] **URL encoding:** Special characters and spaces in state parameters
- [x] **Integration:** Full flow validation with various state configurations

### Manual Testing

- [x] Tested with existing implementation (no state) - works identically
- [x] Tested with custom state parameters - correctly preserved in redirect
- [x] Tested state without `client_id` - automatically appended
- [x] Tested error conditions - proper error handling maintained

### Build & Test Commands Run:

```bash
npm run build        # ✅ Success
npm run test         # ✅ All tests passing (21/21)
npm run lint         # ✅ No errors
```

## Checklist:

- [x] Performed a self-review of my own code
- [x] npm test passes on your machine (21/21 tests passing)
- [x] New tests added or existing tests modified to cover all changes
  - Added 21 comprehensive unit tests
  - 100% coverage of new functionality
  - All edge cases and error scenarios covered
- [x] Code conforms with the style guide
  - Follows existing TypeScript patterns
  - Proper error handling
  - Clear comments explaining logic
- [x] API Documentation in code was updated
  - Inline comments added for all scenarios
  - JSDoc comments maintained
- [x] Any dependent changes have been merged and published in downstream modules
  - No external dependencies required
  - Fully self-contained change

### Security Considerations

- ✅ All existing security validations maintained
- ✅ `client_id` validation still enforced
- ✅ State parameters are properly URL-encoded
- ✅ No sensitive data stored (developer responsibility)
- ✅ CSRF protection maintained through OAuth flow

### Performance Impact

- ✅ Minimal overhead (URLSearchParams parsing)
- ✅ No database or external API calls added
- ✅ Stateless implementation (no server-side storage)

### Use Cases Enabled

1. **Deep Linking:** Return users to their intended page after authentication
2. **Multi-Tenant Apps:** Preserve tenant context across OAuth flow
3. **Session Correlation:** Maintain session state during authentication
4. **Custom Workflows:** Pass application-specific metadata through auth flow

### Breaking Changes

**None.** This is a fully backward-compatible enhancement. All existing implementations continue working without any modifications.

### Migration Required

**None.** Existing projects require zero changes. New functionality is opt-in via the `state` parameter.

---

**Test Results:**

```
  Keycloak Login Controller - Unit Tests
    queryGen function - State Handling
      Scenario 1: No state passed (default behavior)
        ✔ should create state with only client_id when no state is provided in body
        ✔ should create state with only client_id when no state is provided in query
      Scenario 2: State passed without client_id
        ✔ should append client_id to existing state when client_id is missing
        ✔ should handle URL-encoded state parameters without client_id
        ✔ should append client_id when state has only one parameter
      Scenario 3: State passed with client_id and other properties
        ✔ should use state as-is when client_id is already present
        ✔ should preserve all state parameters when client_id is present
        ✔ should not modify state when client_id is at the end
        ✔ should not modify state when client_id is in the middle
    keycloakCallback method - Redirect with State Parameters
      Scenario 1: State with only client_id
        ✔ should redirect with only code parameter when state contains only client_id
      Scenario 2: State with client_id and additional parameters
        ✔ should redirect with code and all additional state parameters
        ✔ should handle multiple state parameters with special characters
      Scenario 3: State with complex query parameters
        ✔ should preserve all state parameters in redirect
      Error Scenarios
        ✔ should throw error when client_id is missing from state
        ✔ should throw error when user is undefined
        ✔ should throw error when client is not found
        ✔ should throw error when client has no redirect URL
        ✔ should log error and throw when getAuthCode fails
      URL Encoding and Special Characters
        ✔ should properly encode state parameters with spaces and special chars
      Integration: Full Flow with Various State Configurations
        ✔ should handle empty string state gracefully with client_id appended
        ✔ should handle state with only non-client_id parameters

  21 passing (59ms)
```

---

**Related Documentation:**

- OAuth 2.0 RFC 6749, Section 4.1.1 (Authorization Request with state parameter)
- OAuth 2.0 Security Best Practices (state parameter usage)

### CODE REVIEW

File: services/authentication-service/src/modules/auth/models/client-auth-request.dto.ts [hunk 1]
```diff
@@ -25,4 +25,10 @@ export class ClientAuthRequest extends CoreModel<ClientAuthRequest> {
     required: true,
   })
   client_secret: string; //NOSONAR
+
+  @property({
+    type: 'string',
+    description: ModelPropertyDescriptionString.reqStrPropDesc,
```
[the remark this answers is not in the document]
Comment (subhamjagurisf) [reply services/authentication-service/src/modules/auth/models/client-auth-request.dto.ts]: fixed
